### PR TITLE
Link Datasource to indexPattern

### PR DESCRIPTION
**Pull request by kristenTian**
### Description
This PR is a POC to link Datasource to indexPattern. The changes include:
- Modifies how index pattern is created, -- also to include datasource as a reference when it exists.
- Add a step in index pattern management to select data source.
 
### Issues Resolved
TBA

### CODE REVIEW

File: src/plugins/data/public/index_patterns/saved_index_patterns/saved_index_patterns.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+interface Services {
+  savedObjectsClient: SavedObjectsClientContract;
+  // indexPatterns: IndexPatternsContract;
+  // search: DataPublicPluginStart['search'];
+  chrome: ChromeStart;
+  overlays: OverlayStart;
+}
+
+export function createSavedIndexPatternLoader(services: Services) {
+  const SavedIndexPattern = createSavedIndexPatternClass(services);
```
Comment (zengyan-amazon) [inline src/plugins/data/public/index_patterns/saved_index_patterns/saved_index_patterns.ts:26]: `camelCase` for variable names

**Comment by noCharger** [comment]
Screen Shot 2022-07-19 at 3 38 26 PM

Can we add a field for data source so that the actual index pattern have selected data source attribute as a reference?

Right now even we select `test` data source, the actual reference is `dataSource_0`

  "references" : [
    {
      "name" : "dataSource_0",
      "type" : "data-source",
      "id" : "280568c0-fa7a-11ec-9be1-d1db20ee5019"
    }
  ],

**Review by noCharger — DISMISSED** [review]
Let's link the datasource to index pattern

https://github.com/elastic-analytics/OpenSearch-Dashboards/pull/17/#issuecomment-1189622826

File: src/plugins/data/common/index_patterns/index_patterns/index_pattern.ts [hunk 2]
```diff
@@ -86,6 +86,8 @@ export class IndexPattern implements IIndexPattern {
   // savedObject version
   public version: string | undefined;
   public sourceFilters?: SourceFilter[];
+  public dataSourcesJSON: string | undefined;
```
Comment (zhongnansu) [inline src/plugins/data/common/index_patterns/index_patterns/index_pattern.ts:89]: - why isn't it `dataSourceId`?
- If it has to be a JSON as the name indicated, why not use the `DataSourceType` you previously defined.

**Review by seraphjiang — APPROVED** [review]

File: src/plugins/data/common/index_patterns/index_patterns/index_pattern.ts (hunk 2 above)
Comment (seraphjiang) [reply src/plugins/data/common/index_patterns/index_patterns/index_pattern.ts:89]: @kristenTian any comments?

**Review by noCharger — APPROVED** [review]
Non-blocking comments

File: src/plugins/data/common/index_patterns/index_patterns/index_pattern.ts (hunk 2 above)
Comment (kristenTian) [reply src/plugins/data/common/index_patterns/index_patterns/index_pattern.ts:89]: offline synced during meeting, will add separate support for get

**Pull request merged**